File: web/src/pipeline/svgio.ts
/**
 * SVG loading and filled-region extraction. Port of ball_stencil/svgio.py.
 *
 * Each <path> carries closed subpaths describing a filled silhouette under the
 * even-odd rule. We flatten every subpath and fold them with even-odd union
 * (== symmetric-difference of all ring interiors) to obtain the filled region,
 * which becomes the holes in the stencil shell.
 */

import type { Paths64 } from "clipper2-js";
import { Clip } from "./clip";
import { ParsedSvg } from "./svg";
import { parsePathSubpaths, flattenSubpath, dedupePolyline, Pt } from "./tessellate";

export interface Artwork {
  region: Paths64; // even-odd filled region, SVG units, snapped to grid
  viewbox: [number, number, number, number];
  center: [number, number];
  labels: string[];
}

function ringArea(pts: Pt[]): number {
  let a = 0;
  const n = pts.length;
  for (let i = 0; i < n; i++) {
    const [x0, y0] = pts[i];
    const [x1, y1] = pts[(i + 1) % n];
    a += x0 * y1 - x1 * y0;
  }
  return a / 2.0;
}

function subpathToRing(sub: ReturnType<typeof parsePathSubpaths>[number], tol: number): number[][] | null {
  let pts = dedupePolyline(flattenSubpath(sub, tol));
  if (pts.length < 3) return null;
  // close the ring
  if (Math.abs(pts[0][0] - pts[pts.length - 1][0]) > 1e-9 || Math.abs(pts[0][1] - pts[pts.length - 1][1]) > 1e-9) {
    pts = [...pts, pts[0]];
  }
  if (Math.abs(ringArea(pts)) <= 0) return null;
  return pts as number[][];
}

export function loadArtwork(
  parsed: ParsedSvg,
  chordTolSvg: number,
  clip: Clip,
  name = "input.svg",
): Artwork {
  const rings: number[][][] = [];
  const labels: string[] = [];
  for (const path of parsed.paths) {
    if (path.hidden) continue;
    labels.push(path.label);
    for (const sub of parsePathSubpaths(path.d)) {
      const ring = subpathToRing(sub, chordTolSvg);
      if (ring) rings.push(ring);
    }
  }
  if (rings.length === 0) {
    throw new Error(`No filled vector paths found in ${name}`);
  }
  // Even-odd fill == XOR of every ring interior, then grid-snapped + cleaned
  // (Shapely set_precision + make_valid). Snapping fixes R_ref to match the
  // reference, which measures the region's extent after set_precision.
  const region = clip.snapToGrid(clip.compact(clip.evenOddUnion(rings)));

  const [minx, miny, w, h] = parsed.viewBox;
  const center: [number, number] = [minx + w / 2.0, miny + h / 2.0];
  return { region, viewbox: parsed.viewBox, center, labels };
}

File: ball_stencil/svgio.py
"""SVG loading and filled-region extraction.

The artwork is *filled* vector shapes (not strokes): each <path> carries many
closed subpaths that together describe a filled silhouette under the even-odd
rule.  We tessellate every subpath and fold them with symmetric-difference to
obtain the exact even-odd filled region as a shapely (Multi)Polygon.

This region is what becomes the *holes* in the stencil shell.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np
import shapely
from shapely.geometry import Polygon
from svgpathtools import svg2paths2

from .tessellate import dedupe_polyline, flatten_path


@dataclass
class Artwork:
    region: shapely.Geometry       # even-odd filled region (Multi)Polygon, SVG units
    viewbox: tuple[float, float, float, float]  # (minx, miny, w, h)
    center: tuple[float, float]    # default mapping centre (viewBox centre)
    labels: list[str]              # path labels, for reference


def _parse_viewbox(svg_attr: dict) -> tuple[float, float, float, float]:
    vb = svg_attr.get("viewBox") or svg_attr.get("viewbox")
    if vb:
        a = [float(x) for x in vb.replace(",", " ").split()]
        return (a[0], a[1], a[2], a[3])
    # fall back to width/height in user units
    w = float(str(svg_attr.get("width", "0")).rstrip("pt").rstrip("px") or 0)
    h = float(str(svg_attr.get("height", "0")).rstrip("pt").rstrip("px") or 0)
    return (0.0, 0.0, w, h)


def load_artwork(svg_path: str, chord_tol_svg: float, snap_grid: float = 0.0) -> Artwork:
    """Load filled paths from ``svg_path`` and return the even-odd region.

    ``chord_tol_svg`` is the Bezier flattening tolerance in SVG user units.
    ``snap_grid`` (>0) grid-snaps the region to clean topology and drop slivers.
    Only <path> elements with a visible fill are used; <image> tags and
    display:none paths are ignored.
    """
    paths, attrs, svg_attr = svg2paths2(svg_path)
    viewbox = _parse_viewbox(svg_attr)

    ring_polys: list[Polygon] = []
    labels: list[str] = []
    for path, attr in zip(paths, attrs):
        style = (attr.get("style") or "")
        if "display:none" in style.replace(" ", "") or attr.get("display") == "none":
            continue
        labels.append(attr.get("inkscape:label") or attr.get("id") or "path")
        for sub in path.continuous_subpaths():
            poly = _subpath_to_polygon(sub, chord_tol_svg)
            if poly is not None:
                ring_polys.append(poly)

    if not ring_polys:
        raise ValueError(f"No filled vector paths found in {svg_path}")

    # Even-odd fill == XOR of every ring interior.
    region = ring_polys[0]
    for poly in ring_polys[1:]:
        region = shapely.symmetric_difference(region, poly)
    region = shapely.make_valid(region)
    if snap_grid > 0:
        region = shapely.make_valid(shapely.set_precision(region, snap_grid))

    minx, miny, w, h = viewbox
    center = (minx + w / 2.0, miny + h / 2.0)
    return Artwork(region=region, viewbox=viewbox, center=center, labels=labels)


def _subpath_to_polygon(sub, chord_tol_svg: float) -> Polygon | None:
    pts = dedupe_polyline(flatten_path(sub, chord_tol_svg))
    if len(pts) < 3:
        return None
    # close the ring
    if np.any(np.abs(pts[0] - pts[-1]) > 1e-9):
        pts = np.vstack([pts, pts[0]])
    poly = Polygon(pts)
    if not poly.is_valid:
        poly = poly.buffer(0)
    if poly.is_empty or poly.area <= 0:
        return None
    return poly
